homeassistant/components/vesync/humidifier.py

```diff
@@ -0,0 +1,201 @@
+
+from .common import is_humidifier
+from .const import (
+    DOMAIN,
+    VS_COORDINATOR,
+    VS_DISCOVERY,
+    VS_HUMIDIFIER_MODE_AUTO,
+    VS_HUMIDIFIER_MODE_HUMIDITY,
+    VS_HUMIDIFIER_MODE_MANUAL,
+    VS_HUMIDIFIER_MODE_SLEEP,
+    VS_HUMIDIFIERS,
+    VeSyncHumidifierDevice,
+)
+from .coordinator import VeSyncDataCoordinator
+from .entity import VeSyncBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+MIN_HUMIDITY = 30
+MAX_HUMIDITY = 80
+
+VS_TO_HA_ATTRIBUTES = {ATTR_HUMIDITY: "current_humidity"}
+
+VS_TO_HA_MODE_MAP = {
+    VS_HUMIDIFIER_MODE_AUTO: MODE_AUTO,
+    VS_HUMIDIFIER_MODE_HUMIDITY: MODE_AUTO,
+    VS_HUMIDIFIER_MODE_MANUAL: MODE_NORMAL,
+    VS_HUMIDIFIER_MODE_SLEEP: MODE_SLEEP,
+}
+
+HA_TO_VS_MODE_MAP = {v: k for k, v in VS_TO_HA_MODE_MAP.items()}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the VeSync humidifier platform."""
+
+    coordinator = hass.data[DOMAIN][VS_COORDINATOR]
+
+    @callback
+    def discover(devices):
+        """Add new devices to platform."""
+        _setup_entities(devices, async_add_entities, coordinator)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, VS_DISCOVERY.format(VS_HUMIDIFIERS), discover)
+    )
+
+    _setup_entities(hass.data[DOMAIN][VS_HUMIDIFIERS], async_add_entities, coordinator)
+
+
+@callback
+def _setup_entities(
+    devices: list[VeSyncBaseDevice],
+    async_add_entities,
+    coordinator: VeSyncDataCoordinator,
+):
+    """Add humidifier entities."""
+    entities = [
+        VeSyncHumidifierHA(dev, coordinator) for dev in devices if is_humidifier(dev)
+    ]
+
+    async_add_entities(entities, update_before_add=True)
+
+
+def _get_ha_mode(vs_mode: str) -> str | None:
+    ha_mode = VS_TO_HA_MODE_MAP.get(vs_mode)
+    if ha_mode is None:
+        _LOGGER.warning("Unknown mode '%s'", vs_mode)
+    return ha_mode
+
+
+def _get_vs_mode(ha_mode: str) -> str | None:
+    vs_mode = HA_TO_VS_MODE_MAP.get(ha_mode)
+    if vs_mode is None:
+        _LOGGER.warning("Unknown mode '%s'", ha_mode)
+    return vs_mode
+
+
+class VeSyncHumidifierHA(VeSyncBaseEntity, HumidifierEntity):
+    """Representation of a VeSync humidifier."""
+
+    # The base VeSyncBaseEntity has _attr_has_entity_name and this is to prevent
+    # none being appended to entity id.
+    _attr_name = ""
+
+    _attr_max_humidity = MAX_HUMIDITY
+    _attr_min_humidity = MIN_HUMIDITY
+    _attr_supported_features: HumidifierEntityFeature = HumidifierEntityFeature(
+        HumidifierEntityFeature.MODES
+    )
+
+    def __init__(
+        self,
+        humidifier: VeSyncHumidifierDevice,
+        coordinator: VeSyncDataCoordinator,
+    ) -> None:
+        """Initialize the VeSync humidifier device."""
+        super().__init__(humidifier, coordinator)
+        self.smarthumidifier = humidifier
```

Comment:
Isn't the device set to the object in the super constructor?

Reply:
It is, the `smarthumidifier` is just better typed thus allowing one to use humidifier specific methods like set_humidity,set_humidity_mode, turn_on and turn_off. I changed the class to define the correct `device` type and removed `smarthumidifier`.